lib/questionlib.php

```diff
@@ -2594,3 +2076,726 @@ function get_next_version(int $questionbankentryid): ?int {
+            $pcontexts[] = $context->id;
+        }
+        $contextslist = join(', ', $pcontexts);
+
+        $categories = get_categories_for_contexts($contextslist, 'qc.parent, qc.sortorder, qc.name ASC', $top);
+
+        if ($top) {
+            $categories = question_fix_top_names($categories);
+        }
+
+        $categories = question_add_context_in_key($categories);
+        $categories = add_indented_names($categories, $nochildrenof);
+
+        // Sort cats out into different contexts.
+        $categoriesarray = array();
+        foreach ($pcontexts as $contextid) {
+            $context = context::instance_by_id($contextid);
+            $contextstring = $context->get_context_name(true, true, $escapecontextnames);
+            foreach ($categories as $category) {
+                if ($category->contextid == $contextid) {
+                    $cid = $category->id;
+                    if ($currentcat != $cid || $currentcat == 0) {
+                        $a = new stdClass;
+                        $a->name = format_string($category->indentedname, true,
+                            array('context' => $context));
+                        if ($category->idnumber !== null && $category->idnumber !== '') {
+                            $a->idnumber = s($category->idnumber);
+                        }
+                        if (!empty($category->questioncount)) {
+                            $a->questioncount = $category->questioncount;
+                        }
+                        if (isset($a->idnumber) && isset($a->questioncount)) {
+                            $formattedname = get_string('categorynamewithidnumberandcount', 'question', $a);
+                        } else if (isset($a->idnumber)) {
+                            $formattedname = get_string('categorynamewithidnumber', 'question', $a);
+                        } else if (isset($a->questioncount)) {
+                            $formattedname = get_string('categorynamewithcount', 'question', $a);
+                        } else {
+                            $formattedname = $a->name;
+                        }
+                        $categoriesarray[$contextstring][$cid] = $formattedname;
+                    }
+                }
+            }
+        }
+        if ($popupform) {
+            $popupcats = array();
+            foreach ($categoriesarray as $contextstring => $optgroup) {
+                $group = array();
+                foreach ($optgroup as $key => $value) {
+                    $key = str_replace($CFG->wwwroot, '', $key);
+                    $group[$key] = $value;
+                }
+                $popupcats[] = array($contextstring => $group);
+            }
+            return $popupcats;
+        } else {
+            return $categoriesarray;
+        }
+    }
+
+    /**
+     * @deprecated since Moodle 4.0 MDL-71585
+     * @see qbank_managecategories\helper
+     */
+    function question_add_context_in_key($categories) {
```

Comment:
remove the deprecated code and point to the new function

Reply:
done, changed to public in managecategories.